index.coffee

```diff
@@ -1,60 +1,159 @@
+			# and emit a more appropriate error.
+			@error = new Error('ResumableRequest is not writable')
+			# have to abort asynchronously to allow client code to register 'error' handlers
+			process.nextTick =>
+				@_retry()
+
+		@_request()
+
+	abort: ->
+		@emit('abort')
+		@destroy()
+
+	push: (data, encoding) ->
+		@retries = 0 # we got some data, reset number of retries
+		@bytesRead += data.length if data?
+		@_reportProgress()
+		return super(data, encoding)
+
+	_destroy: (err, cb) ->
+		return cb(err) if @ended
+		@ended = true
+		doEnd = =>
+			@request.abort()
+			@_retry.cancel()
+			@_reportProgress()
+			@_reportProgress.flush()
+			@_reportProgress.cancel()
+			cb(err)
+			@emit('end')
+			if not @error?
+				@emit('complete')
+		if @response?
+			@response.end(doEnd)
+		else
+			doEnd()
+
+	_read: -> # noop -- we're manually pushing buffers into self
+
+	progress: ->
+		retries: @retries
+		maxRetries: @maxRetries
+		percentage: Math.round(100 * @bytesRead / @bytesTotal) or 0
+		size: { transferred: @bytesRead, total: @bytesTotal }
+
+	_reportProgress: (progress = {}) =>
+		@emit('progress', Object.assign(@progress(), progress))
 
 	# Sends the requests and sets up listeners in order
 	# to be able to resume.
 	# Returns a stream object similar to the result of `request`.
 	# 'response' and 'request' events on the stream are generated
 	# based on the first request, in order to emit them only once.
-	resume: ->
-		@requestOpts.headers ?= {}
-		@requestOpts.headers.range = "bytes=#{@bytesRead}-"
-		@requestModule(@requestOpts)
-		.on 'request', (reqObj) =>
-			if @retries is 0
-				@stream.emit('request', reqObj)
-		.on 'response', (response) =>
+	_request: ->
+		if @response?
+			# the first response is emitted to listeners, so only specify the
+			# Range header on subsequent requests, to avoid causing a 206 response
+			# status code, which may be unexpected.
+			@requestOpts.headers ?= {}
+			@requestOpts.headers.range = "bytes=#{@bytesRead}-"
+
+		initial = @request is null
+		failed = false
+
+		@request = @requestModule(@requestOpts)
+		.once 'request', (reqObj) =>
+			if initial
+				@emit('request', reqObj)
+		.once 'response', (response) =>
+			@_follow(response)
 			if response.statusCode >= 400
 				@error = new Error("Request failed with status code #{response.statusCode}")
-				@stream.emit('error', @error)
-			else if @retries is 0
-				@stream.emit('response', response)
-				@contentLength = parseInt(response.headers['content-length'])
+				@_retry() # will abort the request
 		.on 'error', (err) =>
-			@stream.emit('socketError', err)
-			@retryOrEnd()
-		.on('end', => @retryOrEnd())
-		.pipe(@stream, end: false)
+			failed = true
+			@emit('socketError', err)
+			@_retry()
+		.once 'complete', =>
+			return if failed # we've already triggered a retry
+			if @bytesTotal? and @bytesRead < @bytesTotal
+				@_retry()
+			else
+				@destroy() # received complete response
+
+	_follow: (response) ->
+		if not @response?
```

Comment:
Could you add a comment about the behaviour here, I'm guessing it's along the lines of "grab response info and store it for the first request we make, to use for resume requests later" but you can probably give a much better explanation - it'll help people looking at the code in the future